qiskit_optimization/algorithms/qrao/magic_rounding.py

```diff
@@ -0,0 +1,448 @@
+    basis_counts: List[Dict[str, int]]
+    """The basis_counts represents the resulting counts obtained by measuring with the bases
+    corresponding to the number of shots specified in basis_shots."""
+
+
+class MagicRounding(RoundingScheme):
+    """Magic rounding scheme
+
+    This scheme is described in https://arxiv.org/abs/2111.03167v2.
+    """
+
+    _DECODING = {
+        3: (  # Eq. (8)
+            {"0": [0, 0, 0], "1": [1, 1, 1]},  # I mu+ I, I mu- I
+            {"0": [0, 1, 1], "1": [1, 0, 0]},  # X mu+ X, X mu- X
+            {"0": [1, 0, 1], "1": [0, 1, 0]},  # Y mu+ Y, Y mu- Y
+            {"0": [1, 1, 0], "1": [0, 0, 1]},  # Z mu+ Z, Z mu- Z
+        ),
+        2: (  # Sec. VII
+            {"0": [0, 0], "1": [1, 1]},  # I xi+ I, I xi- I
+            {"0": [0, 1], "1": [1, 0]},  # X xi+ X, X xi- X
+        ),
+        1: ({"0": [0], "1": [1]},),
+    }
+
+    # Pauli op string to label index in ops
+    _OP_INDICES = {1: {"Z": 0}, 2: {"X": 0, "Z": 1}, 3: {"X": 0, "Y": 1, "Z": 2}}
+
+    def __init__(
+        self,
+        sampler: Sampler,
+        basis_sampling: str = "uniform",
+        seed: Optional[int] = None,
+    ):
+        """
+        Args:
+            sampler: Sampler to use for sampling the magic bases.
+            basis_sampling: Method to use for sampling the magic bases.  Must
+                be either ``"uniform"`` (default) or ``"weighted"``.
+                ``"uniform"`` samples all magic bases uniformly, and is the
+                method described in https://arxiv.org/abs/2111.03167v2.
+                ``"weighted"`` attempts to choose bases strategically using the
+                Pauli expectation values from the minimum eigensolver.
+                However, the approximation bounds given in
+                https://arxiv.org/abs/2111.03167v2 apply only to ``"uniform"``
+                sampling.
+            seed: Seed for random number generator, which is used to sample the
+                magic bases.
+
+        Raises:
+            ValueError: If ``basis_sampling`` is not ``"uniform"`` or ``"weighted"``.
+        """
+        if basis_sampling not in ("uniform", "weighted"):
+            raise ValueError(
+                f"'{basis_sampling}' is not an implemented sampling method. "
+                "Please choose either 'uniform' or 'weighted'."
+            )
+        self._sampler = sampler
+        self.rng = np.random.RandomState(seed)
+        self._basis_sampling = basis_sampling
+        self._shots = None
+        super().__init__()
+
+    @property
+    def sampler(self) -> Sampler:
+        """Returns the ``Sampler`` used to sample the magic bases."""
+        return self._sampler
+
+    @property
+    def basis_sampling(self):
+        """Basis sampling method (either ``"uniform"`` or ``"weighted"``)."""
+        return self._basis_sampling
+
+    @staticmethod
+    def _make_circuits(
+        circuit: QuantumCircuit, bases: np.ndarray[Any, Any], vars_per_qubit: int
```

Comment:
```suggestion
        circuit: QuantumCircuit, bases: np.ndarray, vars_per_qubit: int
```

See https://github.com/Qiskit/qiskit-optimization/actions/runs/4956834568/jobs/8867763382?pr=487#step:7:447